HarfangLab/harfanglab/download_file_from_endpoint.py

```diff
@@ -0,0 +1,155 @@
+# coding: utf-8
+import hashlib
+import io
+import pathlib
+import shutil
+import tempfile
+import urllib.parse
+import uuid
+from typing import Any, TypeAlias
+
+import requests
+import tenacity
+from harfanglab.job_executor import JobExecutor
+from harfanglab.models import JobAction, JobTarget
+from tenacity import retry_if_exception_message, stop_after_delay, wait_exponential
+
+
+StrOrUUID4: TypeAlias = str | uuid.UUID
+
+
+class DownloadFileFromEndpointAction(JobExecutor):
+    """
+    Action to download a single file from an HarfangLab endpoint
+    """
+
+    _artefact_info: dict[str, Any] | None = None
+    _artefact_info_fetched: bool = False
+
+    @property
+    def artefact_sha256(self) -> str:
+        if not self._artefact_info_fetched:
+            raise RuntimeError("DownloadFileFromEndpointAction.fetch_artefact_info() not called")
+        if self._artefact_info is None:
+            # Note: This raise is mostly for semantic/mypy purpose and should never
+            # happen in real life (an error will always raise before)
+            raise RuntimeError(
+                "DownloadFileFromEndpointAction.fetch_artefact_info() has been called, "
+                "but '_artifact_info' is still None - Something really strange and wrong happen"
+            )
+        return self._artefact_info["sha256"]
```

Comment:
When the file is not found, the `sha256` is undefined. Can you handle the case when the path to the file is incorrect or the download of the file failed?
Thanks

Reply:
Sample of the job response when the file is not found:

```json
{
  "count": 1,
  "next": null,
  "previous": null,
  "results": [
    {
      "shortname": "C:\\Users\\<REDACTED>\\Downloads\\hello.txt",
      "artefact_type": "downloaded file",
      "item_status": 0,
      "download_status": 1,
      "destination": "elastic",
      "job_id": "<REDACTED>",
      "@version": "1",
      "tenant": "<REDACTED>",
      "log_type": "investigation",
      "msg": "0 file(s) (0B) downloaded",
      "@timestamp": "2024-10-17T15:13:43.619423Z",
      "size": 0,
      "agent": {
        "ostype": "windows",
        "version": "4.1.6",
        "osversion": "10.0.22621",
        "osproducttype": "Windows 11 Pro",
        "domainname": "WORKGROUP",
        "agentid": "<REDACTED>",
        "hostname": "Windows11"
      },
      "date": "2024-10-17T15:13:42.975214Z",
      "job_instance_id": "<REDACTED>",
      "id": "<REDACTED>"
    }
  ],
  "filters": {
    "wildcard": [
      "artefact_type",
      "msg",
      "shortname",
      "longname",
      "agent.domainname",
      "agent.hostname",
      "agent.osproducttype"
    ]
  }
}
```